### PR TITLE
feat(tensor): add validated column-major host views

### LINKED ISSUE #1736: [Feature]: Add validated const-generic column-major host views for zero-overhead scalar indexing
## What are you trying to do?

Write ordinary Julia-/Fortran-style scalar loops over host-resident, compact column-major `TypedTensor<T, Rank<N>>` values while retaining multidimensional shape information and allowing release optimization to remove per-element bounds, rank, layout, and backend checks.

The intended usage should support parametric rank rather than a growing `get2`/`get3` family, for example:

```rust
let a = tensor.host_col_major()?;

for k in 0..a.shape()[2] {
    for j in 0..a.shape()[1] {
        for i in 0..a.shape()[0] {
            sum += a[[i, j, k]];
        }
    }
}
```

There should also be an iteration form whose absence of per-element bounds checks follows from the API structure rather than depending only on LLVM proving that repeated checked indexing is redundant.

## What makes it hard today?

`TypedTensor` is compact column-major and already exposes `as_slice()`, `iter()`, and `iter_mut()`, but those APIs reduce access to a flat buffer. Multidimensional access currently has several hot-loop costs:

- `get(&[usize])` and `get2`/`get3` validate rank and bounds and return `Result` on each access.
- `get_unchecked(&[usize])` still accepts a dynamic index slice, recomputes the mixed-radix offset through a dynamic loop, performs checked offset arithmetic, and calls `host_data()?` per access.
- `TypedTensorView::as_slice()` can prove that a view is contiguous, but the result is only a flat slice; the validated shape/layout state is not represented by a reusable public host-view type.
- Dropping to `strided-kernel` is inappropriate for an ordinary scalar loop over already-contiguous host memory. That layer should remain for bulk kernels and arbitrary-strided execution.

Relevant current code at commit `8b768af144057b3d25ce70524b34e0570b38ef85`:

- [element accessors](https://github.com/tensor4all/tenferro-rs/blob/8b768af144057b3d25ce70524b34e0570b38ef85/crates/tenferro-tensor/src/types/accessors.rs)
- [`TypedTensorView` layout and contiguous-slice APIs](https://github.com/tensor4all/tenferro-rs/blob/8b768af144057b3d25ce70524b34e0570b38ef85/crates/tenferro-tensor/src/types.rs)
- [existing element-access benchmarks](https://github.com/tensor4all/tenferro-rs/blob/8b768af144057b3d25ce70524b34e0570b38ef85/crates/tenferro-tensor/benches/element_access.rs)

## Proposed behavior or API

The exact names are open for design review. The desired contract is a validated, const-generic, host-resident column-major view:

```rust
pub struct ColMajorView<'a, T, const N: usize> {
    // Private fields. Construction proves the invariants below.
}

pub struct ColMajorViewMut<'a, T, const N: usize> {
    // Private fields with exclusive, injective access.
}
```

Possible constructors:

```rust
impl<T, const N: usize> TypedTensor<T, Rank<N>> {
    pub fn host_col_major(&self) -> Result<ColMajorView<'_, T, N>>;
    pub fn host_col_major_mut(&mut self) -> Result<ColMajorViewMut<'_, T, N>>;
}

impl<'a, T, const N: usize> TypedTensorView<'a, T, Rank<N>> {
    pub fn host_col_major(&self) -> Result<ColMajorView<'_, T, N>>;
}
```

Construction should check once, before the hot loop:

- host residency, with no implicit device download;
- rank `N`;
- checked shape product;
- compact column-major strides;
- valid offset and reachable backing range;
- exact logical slice length;
- injectivity/exclusive borrowing for mutable views.

The resulting type should expose rank-parametric access rather than requiring rank-specific implementations:

```rust
impl<T, const N: usize> ColMajorView<'_, T, N> {
    pub fn shape(&self) -> &[usize; N];
    pub fn as_slice(&self) -> &[T];

    pub fn get(&self, index: [usize; N]) -> Option<&T>;

    /// Caller proves index[axis] < shape[axis] for every axis.
    pub unsafe fn get_unchecked(&self, index: [usize; N]) -> &T;
}

impl<T, const N: usize> Index<[usize; N]> for ColMajorView<'_, T, N> {
    type Output = T;
}
```

Because `N` is a compile-time constant, an inlined offset loop can be unrolled. Checked construction should also prove that ordinary unchecked `usize` offset arithmetic cannot overflow for an in-bounds index. A leaf unchecked accessor must not repeat host/backend checks or return `Result`.

Safe traversal APIs should be the primary guaranteed fast path:

```rust
for value in a.iter() {
    // linear column-major order
}

for lane in a.axis0_lanes() {
    for value in lane {
        // first index varies fastest, as in Fortran and Julia
    }
}
```

For mutable traversal, safe APIs must yield disjoint references by construction, such as `iter_mut()` and `axis0_lanes_mut()`. A safe API must not create multiple mutable references from arbitrary repeated indices.

A dynamic-rank counterpart may be considered separately, but it must not weaken or complicate the `Rank<N>` fast path.

## Comparison with Julia and Fortran

Julia dense arrays are column-major, and its performance guidance recommends varying the first index in the innermost loop. Julia's array interface also separates indexing policy from traversal:

- `IndexStyle` identifies efficient linear versus Cartesian indexing.
- `eachindex` returns an efficient iteration domain for the array type.
- `getindex` may place checks in an `@boundscheck` block.
- A caller can use `@inbounds`; after inlining, Julia may elide the marked bounds checks.

References:

- [Julia bounds-checking internals: `@inbounds`, `@boundscheck`, and inlining](https://docs.julialang.org/en/v1/devdocs/boundscheck/)
- [Julia `eachindex` and `IndexStyle`](https://docs.julialang.org/en/v1/base/arrays/#Base.eachindex)
- [Julia performance guidance for column-major access](https://docs.julialang.org/en/v1/manual/performance-tips/#Access-arrays-in-memory-order,-along-columns)

The proposed Rust API should preserve Rust's explicit safety boundary:

- validated construction plus safe iterators corresponds to an efficient `eachindex`-style traversal whose valid domain is encoded by the API;
- checked `get([usize; N])` provides ordinary safe indexing;
- `unsafe get_unchecked([usize; N])` is the explicit leaf-kernel analogue of Julia's `@inbounds`, with a documented proof obligation;
- unlike a broad annotation or macro, unchecked access should not propagate implicitly through unrelated call layers.

Fortran dense arrays likewise use column-major addressing, so a loop with the first subscript innermost can reduce to a base pointer plus incremental offsets once extents are established. tenferro should make that loop structure expressible without per-element `Result`, backend dispatch, dynamic-rank iteration, or layout validation.

## Relevant area

- `tenferro-tensor`
- `TypedTensor<T, Rank<N>>`
- compact host `TypedTensorView<T, Rank<N>>`
- public API, documentation, safety invariants, and performance benchmarks

## Dependency, backend, or AD impact

- No new dependency should be necessary.
- This is a host-access API. CUDA/WebGPU/backend-owned storage must return a typed error before entering the loop; no implicit transfer or synchronization should be introduced.
- Arbitrary-strided backend execution remains owned by existing backend/kernel abstractions.
- No change to eager/traced semantics or AD behavior is intended.
- No new feature flag is expected.
- Mutable APIs must preserve Rust aliasing and lifecycle soundness.

## Non-goals

- Replacing `strided-kernel` for arbitrary-strided, broadcast, fused, or parallel tensor kernels.
- Making GPU device memory directly indexable from host Rust.
- Guaranteeing zero checks for arbitrary safe random indices that are not derived from a validated domain.
- Adding separate primary implementations for every rank (`get2`, `get3`, ...).
- Introducing implicit materialization, canonicalization, or device download.
- Changing column-major layout semantics.

## Acceptance criteria

- [ ] A public API converts host-resident compact `TypedTensor<T, Rank<N>>` into a const-generic validated column-major view after one boundary validation.
- [ ] A compact offset `TypedTensorView<T, Rank<N>>` can use the same abstraction without copying; noncompact or backend views return typed errors.
- [ ] Shape is exposed as `[usize; N]`, and multidimensional access accepts `[usize; N]` rather than a dynamic index slice on the static-rank path.
- [ ] Checked scalar access is available for convenience.
- [ ] A documented unsafe scalar accessor performs no per-access rank, bounds, backend, layout, or `Result` work in release builds.
- [ ] Safe linear and first-axis-lane iterators perform no per-element bounds check and preserve column-major logical order.
- [ ] Mutable iterators prove disjointness and do not permit safe aliasing through repeated logical indices.
- [ ] Empty dimensions, scalar/rank-0 tensors, singleton dimensions, nonzero compact offsets, shape-product overflow, invalid backing ranges, and backend storage have focused tests.
- [ ] Rustdoc includes const-generic rank-2 and rank-3 examples and explicitly explains that the first index varies fastest.
- [ ] Existing `Tensor`, eager, traced, GPU, AD, and arbitrary-strided behavior remains unchanged.

## Performance and verification

Follow the repository's performance-gated experiment protocol before implementation. Record the baseline/candidate commits, compiler/toolchain, target CPU, profile, affinity/thread settings, cases, repetitions, noise gates, and thresholds before running the candidate.

Extend `crates/tenferro-tensor/benches/element_access.rs` with at least:

- rank 2 and rank 3 canonical Fortran-order loops;
- raw host slice/manual column-major addressing baseline;
- validated-view safe indexing;
- validated-view safe `axis0_lanes()`;
- validated-view unsafe indexing;
- mutable traversal;
- a representative nontrivial scalar loop that can vectorize.

Predeclare a primary gate in the experiment record. A suggested starting gate is:

- safe `axis0_lanes()` median time no more than 5% slower than the equivalent raw-slice loop for each primary rank-2/rank-3 case;
- no statistically significant regression in existing direct-slice and tensor-iterator cases.

Inspect optimized assembly or LLVM IR for representative rank-2/rank-3 probes. The inner loop for the guaranteed fast iterator path must contain no conditional branch attributable to rank, bounds, backend, layout, or `Result` checks. Verify that expected vectorization is not inhibited relative to the raw-slice baseline.

Any internal `unsafe` must have a local `// SAFETY:` proof tied to the checked constructor and boundary tests. Public safe access must remain memory-safe even if LLVM performs no bounds-check elimination.

## Provenance

This issue contains no prototype or third-party source code. Julia is used only as an API/performance comparison; implementation should be original and follow tenferro's `MIT OR Apache-2.0` licensing.

### PR DESCRIPTION
**Pull request by terasakisatoshi**
## Type

- [ ] Bug fix
- [ ] Documentation
- [ ] Refactor / cleanup
- [x] Implementation of accepted issue

## Related issue

Fixes #1736

## Summary

- Add const-generic `ColMajorView` and `ColMajorViewMut` wrappers for validated,
  zero-copy access to compact host-resident tensors.
- Validate backend placement, column-major compactness, shape products, offsets,
  and exact logical slice lengths once at the API boundary.
- Provide recoverable checked `get`/`get_mut` access with `[usize; N]`, safe
  `axis0_lanes()` traversal for the guaranteed fast path, and documented unsafe
  leaf access for callers that can prove bounds.
- Add mutable disjoint traversal, compact offset-view support, focused edge-case
  and Poisson Jacobi tests, Rustdoc examples, design documentation, codegen
  probes, and rank-2/rank-3 Criterion cases.

Checked scalar access now looks like:

```rust
let a = tensor.host_col_major_view()?;

for k in 0..a.shape()[2] {
    for j in 0..a.shape()[1] {
        for i in 0..a.shape()[0] {
            if let Some(value) = a.get([i, j, k]) {
                sum += *value;
            }
        }
    }
}
```

The views deliberately do not implement `Index` or `IndexMut`, because those
traits cannot return a recoverable out-of-bounds result and public invalid input
must not become a panic.

For traversal whose lack of per-element multidimensional bounds checks follows
from the API structure:

```rust
for lane in a.axis0_lanes() {
    for value in lane {
        sum += *value;
    }
}
```

### Corresponding Julia loops

For a matrix, the closest Julia spelling is `eachcol`, because each returned
column is contiguous along the first axis:

```julia
for lane in eachcol(a)
    for value in lane
        sum += value
    end
end
```

Written with explicit indices, the column-major loop keeps the first index in
the innermost loop:

```julia
for j in axes(a, 2)
    for i in axes(a, 1)
        sum += a[i, j]
    end
end
```

The same ordering extends to rank 3 by placing `i` inside `j` inside `k`.
For a rank-independent whole-array reduction, the Rust lane loop is also
equivalent in traversal order to Julia's linear iteration:

```julia
for value in a
    sum += value
end
```

Both traverse dense column-major storage with the first coordinate varying
fastest. `axis0_lanes()` makes that contiguous domain explicit as Rust slices,
so the safe inner loop does not perform multidimensional offset calculation or
per-element multidimensional bounds checks.

The declared performance gate passed: safe axis-0 lanes were 0.98% slower than
the raw-slice median for rank 2 and 1.35% faster for rank 3. Existing iterator
benchmarks had no regression above the 5% threshold.

## Work log

- [`docs/worklogs/2026-08-31-issue-1736-col-major-host-view-poc.md`](docs/worklogs/2026-08-31-issue-1736-col-major-host-view-poc.md)

## Verification

- `cargo test -p tenferro-tensor` (274 unit tests, integration/UI tests, and 352 doctests)
- `cargo test -p tenferro-tensor types::col_major::tests` (8 focused tests)
- `cargo test -p tenferro-tensor --doc types::col_major` (23 focused doctests)
- `cargo clippy -p tenferro-tensor --all-targets -- -D warnings`
- `bash scripts/check-pr-fast.sh --coverage-reviewed --test 'cargo test -p tenferro-tensor types::col_major::tests'`
- deterministic repository-rules review: pass
- Criterion release benchmarks using the protocol recorded in the work log

The static-rank codegen checker is inconclusive locally because its label parser
expects ELF labels while the generated Apple assembly uses Mach-O `LBB...`
labels. Manual inspection found no rank, bounds, backend, layout, allocation, or
`Result` handling inside the representative read/write loops. Hosted Linux
validation should provide the supported automated result. The complete docs-site
check also requires Python 3.11+, while its local snippet phase passed.

## Checklist

- [x] I read `CONTRIBUTING.md`.
- [x] I added or updated tests/docs as needed.
- [x] I ran local repository-rules review (deterministic; the external LLM
      review is permanently disabled) and resolved or documented its findings.
- [x] If this implements a new feature, the linked issue has been accepted by
      maintainers.
- [x] If this is an AI-assisted bug fix, I applied the bug-fix scope gate from
      `ai/contribution-workflows/bugfix-pr.md` (not applicable: accepted feature).
- [x] If this PR needs a work log, I added one under `docs/worklogs/` and
      linked it above.

Generated with [OpenAI Codex](https://openai.com/codex/)

### CODE REVIEW

**Comment by terasakisatoshi** [comment]
## Claude Fable review

Fable reviewed the complete `origin/main...HEAD` diff with emphasis on unsafe soundness, mutable aliasing, shape arithmetic, compact offset views, backend/layout validation, API contracts, tests, and benchmarks. It reported **no blocking findings** and confirmed the unsafe proof chain and iterator disjointness.

Follow-up commit `92eca1e5` addresses the actionable minor findings:

- return the shared validated view with the backing-data lifetime;
- reuse the existing checked view element-count helper;
- remove a duplicate compactness check on `TypedTensorView`;
- cover shared-view `Debug` and the loosened lifetime;
- document the deferred `TypedTensorViewMut` constructor.

A subsequent hosted inline review identified that public `Index`/`IndexMut`
would turn invalid coordinates into a panic, contrary to the repository's
public-boundary contract. Commit `6d5ecf3b` removes those trait implementations
and retains recoverable `get`/`get_mut` as the checked random-access surface.

Deferred follow-ups are `IntoIterator`/`Copy`/`Clone` ergonomics, dynamic-rank access, and the mutable tensor-view constructor. After the changes, focused tests, 23 focused doctests, Clippy with `-D warnings`, public-error-doc checks, and deterministic repository-rules review pass.

File: crates/tenferro-tensor/src/types/col_major.rs [hunk 1]
```diff
@@ -0,0 +1,601 @@
...
+}
+
+impl<T, const N: usize> Index<[usize; N]> for ColMajorView<'_, T, N> {
+    type Output = T;
+
+    #[inline(always)]
+    fn index(&self, index: [usize; N]) -> &Self::Output {
+        match self.get(index) {
+            Some(value) => value,
+            None => panic!("index {index:?} out of bounds for shape {:?}", self.shape),
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply crates/tenferro-tensor/src/types/col_major.rs]: Addressed in commit 6d5ecf3b. I removed the public Index and IndexMut implementations and retained get/get_mut as the recoverable checked random-access surface. Rustdoc, tests, benchmarks, the design note, worklog, and PR body now use the same contract. Focused tests, 23 focused doctests, Clippy with -D warnings, and the public-error-doc check pass locally.

File: crates/tenferro-tensor/src/types/col_major.rs [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+    /// ```rust
+    /// use tenferro_tensor::{Rank, TypedTensor};
+    /// let tensor = TypedTensor::<i32, Rank<1>>::from_vec_col_major([2], vec![4, 5])?;
+    /// let view = tensor.host_col_major()?;
+    /// // SAFETY: index 1 is below the only axis extent, 2.
+    /// assert_eq!(unsafe { view.get_unchecked([1]) }, &5);
+    /// # Ok::<(), tenferro_tensor::Error>(())
+    /// ```
+    #[inline(always)]
+    pub unsafe fn get_unchecked(&self, index: [usize; N]) -> &T {
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply crates/tenferro-tensor/src/types/col_major.rs:182]: Addressed in commit fe7ec6a9. Each of the three production unchecked slice-access blocks now has a nearby canonical INVARIANT marker naming ColMajorView::new or ColMajorViewMut::new and the validate_slice_len checked-product/exact-length proof. Clippy with -D warnings, all 8 focused tests, formatting, and git diff checks pass locally.

File: crates/tenferro-tensor/src/types/col_major.rs [hunk 3]
```diff
@@ -0,0 +1,556 @@
...
+use std::fmt;
+
+use tenferro_tensor_core::ValidationError;
+
+use super::{checked_view_element_count, Rank, TensorScalar, TypedTensor, TypedTensorView};
+
+fn static_shape<const N: usize>(shape: &[usize], op: &'static str) -> crate::Result<[usize; N]> {
+    shape
+        .try_into()
+        .map_err(|_| crate::Error::rank_mismatch(op, N, shape.len()))
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply crates/tenferro-tensor/src/types/col_major.rs]: Addressed in commit c1a04040. The synthetic static_shape conversion and its unreachable rank-mismatch test are removed. Static-rank constructors now obtain the existing array directly through TensorLayout<Rank<N>>::shape_array(), and their error docs no longer claim an observable rank failure. Core/tensor Clippy, focused tests, and both affected doctest sets pass locally.

File: crates/tenferro-tensor/src/types/col_major.rs [hunk 4]
```diff
@@ -0,0 +1,556 @@
...
+    /// assert_eq!(tensor.host_col_major()?.get([1, 0]), Some(&2));
+    /// # Ok::<(), tenferro_tensor::Error>(())
+    /// ```
+    ///
+    /// # Errors
+    ///
+    /// Returns [`crate::Error::RuntimeState`] when storage is backend-owned,
+    /// or [`crate::Error::Validation`] when rank, compact layout, shape
+    /// arithmetic, or the logical host range is invalid.
+    pub fn host_col_major(&self) -> crate::Result<ColMajorView<'_, T, N>> {
```
[the remark this answers is not in the document]
Comment (terasakisatoshi) [reply crates/tenferro-tensor/src/types/col_major.rs]: Addressed in commit c1a04040. The zero-copy metadata accessor family is now named host_col_major_view(), host_col_major_view_mut(), and TypedTensorView::host_col_major_view(). Rustdoc, tests, benchmarks, design/worklog documentation, and the PR example were updated consistently. Core/tensor Clippy and affected doctests pass locally.
